### ISSUE TITLE
Option for not printing full `expected` and `actual` in mismatch message

### ISSUE THREAD

**Issue by philomates** [nubank/matcher-combinators#159]
Currently when a mismatch happens (in clojure.test) we use the entire `(match? the-expected the-actual)` form as the `Expected: ` field and the annotated mismatch diff as the `Actual: ` field. This means that if `the-expected` or `the-actual` matcher/data-structures are large, we get really big output.

In some cases, like when you are matching and expected file contents with an actual file contents, it really suffices to only see the precise mismatch info. This is partially the motivation behind https://github.com/nubank/matcher-combinators/pull/157.

What would be useful to look into is a way to abbreviate or omit aspects of the output.

**Comment by philomates** [comment]
one suggestion by @gusbicalho is an annotation construct:

```clojure
(match?
  (-> expected (reported-as "contents of foo.graphql"))
  (-> actual (reported-as "contents of tmp-file.graphql")))
```

**Comment by philomates** [comment]
another option would be to limit the size of the `Expected: ` output at print time. Something like https://github.com/nubank/matcher-combinators/compare/expected-printing?expand=1# which would for
```clojure
(is (match? [1 (m/not-matcher odd?) 3]
            [1 2 3]))
```
output
```
FAIL in (absent-pred) (test_test.clj:120)
expected: (match? [1 ...)
  actual: [1
 (mismatch
  "expected mismatch from: clojure.core$even_QMARK_@6a0a7d89"
  2)
 3]
```

**Comment by gabrielgiussi** [comment]
Couldn't we just traverse the result and redact/remove anything that is not a Mismatch?
If we run this
```clojure
(is (match? {:a {:b odd?}}
              {:a {:b 2
                   :c 3}}))
```

We have this result
```clojure
{:matcher-combinators.result/type :mismatch,
 :matcher-combinators.result/value {:a {:b #matcher_combinators.model.Mismatch{:expected (pred
                                                                                          #object[matcher_combinators.matchers$pred
                                                                                                  0x46414468
                                                                                                  "matcher_combinators.matchers$pred@46414468"]),
                                                                               :actual 2},
                                        :c 3}},
 :matcher-combinators.result/weight 1}
```

I would like to see
```clojure
  actual: {:a
 {:b
  (mismatch
   (expected
    (pred
          #object[matcher_combinators.matchers$pred 0x46414468 "matcher_combinators.matchers$pred@46414468"]))
   (actual 2))}}
; c was removed completely
```

For other collections where the order is important we could have something like 
```clojure
  actual: ['match (mismatch (expected 2) (actual 1)) 'match]
```

I didn't consider yet if this will play nicely with all the matchers (prefix, in-any-order, etc).
This kind be done as part of the pprint logic, so we should run a filter over the mismatch result before calling to pprint, for example if a *print-summary* var is set to true.

**Comment by gabrielgiussi** [comment]
> I guess we could collect this data in the initial ascent from the recursive descent that computes mismatches and assoc it to the metadata of container datastructures (maps, vectors, etc) and then use metadata in a prewalk to omit any node that has a mismatching subnode.

I can give it a try

**Comment by gabrielgiussi** [comment]
@philomates could you provide me some feedback on [this approach]((https://github.com/nubank/matcher-combinators/pull/209)) before I move further with this implementation? 🙏 

I'm not happy having to introduce the Match wrapper, besides making tests harder as it shows in this PR it might get in the middle of the validation itself which is something I've tried to avoid, but my initial attempt of using meta fell short for values that don't implement IObj which can be associated with metadata, in this category we have numbers, strings and keywords. However if this impl seems too intrusive we could take a best effort approach and only drop of redact data structures like maps, vectors, lists and sets which at the end are the ones which might be flooding the console due to its size. However at first, having an output like the following seems confusing
```clojure
['matched 1 (mismatch (expected 2) (actual 3))]
```
Why the first was redacted to `matched` but the second was kept as 1?

**Comment by gabrielgiussi** [comment]
Actually this provides some value already because removes expected matched keys, it is not able to remove unmatched but unexpected keys, aka those keys that are in the actual value but which are not in the expected value, which is a pretty common case.
```clojure
; the current impl won't be able to filter :a from the output, only :b because is wrapped in a Matched
(is (match? {:a 1 :b 2} {:b 2 :c 3}))
```
I played around with assuming everything is matched if is not one of the wrappers in model.clj but I had some troubles with this approach although I can't remember, I'll try again.

**Comment by philomates** [comment]
Here is a sketch of what I was thinking: 
https://github.com/nubank/matcher-combinators/compare/abbreviated-printing?expand=1

It seems to work in the limited cases that I tested

I guess either one works; feels like introducing the `Match` record is more explicit. But for the ellision/abbreviation implementation I think we only care about container datastructures (maps and sequentials), so it does feel like a lot of wrapping. I suppose the wrapping isn't necessarily bad unless it leaks into un-pretty-printed results.

**Comment by gabrielgiussi** [comment]
Yes, I've started exactly with that (although the mismatch-map is a cleaver way to not having to ask for `map?` in the prewalk, I wasn't using neither of those, nice) and explore another alternative just because not filtering scalar values seemed a confusing output, but I think I prefer the simplicity of your change at the cost of not filtering everything, because I also think that _for the ellision/abbreviation implementation I think we only care about container datastructures_.
Besides, the user will be turning on this behavior explicitly so we can just document that scalar values aren't filtered out. For this wdyt about the dynamic var?

```clojure
(binding [*print-summary* true]
    (is (match? {:name/first "Alfredo"
                 :f [1 2 3]}
                {:name/first  "Afredo"
                 :f [3 2 1]
                 :name/last   "da Rocha Viana"
                 :name/suffix "Jr."})))
```

**Comment by philomates** [comment]
The thing I'm a bit hung up on is the ergonomics of this. Ideally, if I want to use the feature, I shouldn't have to change source forms one by one. For example, I might really like it and want it on be default everywhere. We could prototype with a dynamic var, but I think it is important to figure out some sort of global config system that allows controlling this feature before expecting anyone to use it. 

What do you think?

I'm not familiar with what patterns can be used for setting configurations in a clojure library. Is it possible via deps.edn somehow? Environment variables are an option, in my view they are quick but messy.

**Comment by philomates** [comment]
released an approach in 3.8.7 that can be enabled by calling `(matcher-combinators.config/enable-abbreviation!)`
